src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/FallbackProjectManager.cs

```diff
@@ -0,0 +1,163 @@
+    private readonly LanguageServerFeatureOptions _languageServerFeatureOptions;
+    private readonly ProjectSnapshotManagerAccessor _projectSnapshotManagerAccessor;
+
+    private ImmutableHashSet<ProjectId> _fallbackProjectIds = ImmutableHashSet<ProjectId>.Empty;
+
+    [ImportingConstructor]
+    public FallbackProjectManager(
+        ProjectConfigurationFilePathStore projectConfigurationFilePathStore,
+        LanguageServerFeatureOptions languageServerFeatureOptions,
+        ProjectSnapshotManagerAccessor projectSnapshotManagerAccessor)
+    {
+        _projectConfigurationFilePathStore = projectConfigurationFilePathStore;
+        _languageServerFeatureOptions = languageServerFeatureOptions;
+        _projectSnapshotManagerAccessor = projectSnapshotManagerAccessor;
+    }
+
+    internal void DynamicFileAdded(ProjectId projectId, ProjectKey razorProjectKey, string projectFilePath, string filePath)
+    {
+        if (_fallbackProjectIds.Contains(projectId))
+        {
+            // If this is a fallback project, then Roslyn may not track documents in the project, so these dynamic file notifications
+            // are the only way to know about files in the project.
+            AddFallbackDocument(razorProjectKey, filePath, projectFilePath);
+        }
+        else if (_projectSnapshotManagerAccessor.Instance.GetLoadedProject(razorProjectKey) is null)
+        {
+            // We have been asked to provide dynamic file info, which means there is a .razor or .cshtml file in the project
+            // but for some reason our project system doesn't know about the project. In these cases (often when people don't
+            // use the Razor or Web SDK) we spin up a fallback experience for them
+            AddFallbackProject(projectId, filePath);
+        }
+    }
+
+    internal void DynamicFileRemoved(ProjectId projectId, string projectFilePath, string filePath)
+    {
+        if (_fallbackProjectIds.Contains(projectId))
+        {
+            // If this is a fallback project, then Roslyn may not track documents in the project, so these dynamic file notifications
+            // are the only way to know about files in the project.
+            RemoveFallbackDocument(projectId, filePath, projectFilePath);
+        }
+    }
+
+    private void AddFallbackProject(ProjectId projectId, string filePath)
+    {
+        var project = TryFindProjectForProjectId(projectId);
+        if (project?.FilePath is null)
+        {
+            return;
+        }
+
+        var intermediateOutputPath = Path.GetDirectoryName(project.CompilationOutputInfo.AssemblyPath);
+        if (intermediateOutputPath is null)
+        {
+            return;
+        }
+
+        if (!ImmutableInterlocked.Update(ref _fallbackProjectIds, (set, id) => set.Add(id), project.Id))
+        {
+            return;
+        }
+
+        var rootNamespace = project.DefaultNamespace ?? "ASP";
```

Comment:
Is "ASP" the canonical name when a project doesn't have a default namespace?

Reply:
Yeah, this mirrors what the source generator does: https://github.com/dotnet/razor/blob/main/src/Compiler/Microsoft.NET.Sdk.Razor.SourceGenerators/RazorSourceGenerator.RazorProviders.cs#L51